YUM4/DNF compatibility via yum action plugin

maxamillion opened the pull request:
Signed-off-by: Adam Miller <admiller@redhat.com>

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
In the past, there was only yum which is now known as yum3 or yum 3.x but now there is a "new yum" known as yum4 or yum 4.x which is actually just a yum command-line compatibility layer on top of dnf and since the Ansible modules for yum(aka yum3) and dnf(aka yum4) call each of yum3 and yum4's python APIs natively on the backend, we need to handle this here and pass off to the correct Ansible module to execute on the remote system


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request


##### COMPONENT NAME
<!--- Name of the module, plugin, task or feature -->
yum

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (plugins/action/yum 6ebfce00a0) last updated 2018/08/14 15:51:55 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/admiller/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/admiller/src/dev/ansible/lib/ansible
  executable location = /home/admiller/src/dev/ansible/bin/ansible
  python version = 3.6.6 (default, Jul 19 2018, 14:25:17) [GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]
```

maxamillion committed:
YUM4/DNF compatibility via yum action plugin

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
fix copyright

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
incorporate abadger's refactor to remove AnsibleAction*

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
increase yum/dnf compatibility

DNF does not natively support allow_downgrade as an option, instead
that is always the default (not configurable by the administrator)
so it had to be implemented

 - Fixed group actions in check mode to report correct changed state
 - Better error handling for depsolve and transaction errors in DNF
 - Fixed group action idempotent transactions
 - Add use_backend to yum module/action plugin
 - Fix dnf handling of autoremove (didn't used to work nor had a
   default value specified, now does work and matches default
   behavior of yum)
 - Enable installroot tests for yum4(dnf) integration testing, dnf
   backend now supports that
 - Switch from zip to bc for certain package install/remove test
   cases in yum integration tests. The dnf depsolver downgrades
   python when you uninstall zip which alters the test environment
   and we have no control over that.
 - Add changelog fragment

Signed-off-by: Adam Miller <admiller@redhat.com>

webknjaz commented on lib/ansible/modules/packaging/os/dnf.py:
```diff
@@ -287,14 +287,20 @@ def fetch_rpm_from_url(self, spec):
-                self.module.fail_json(msg="Failure downloading %s, %s" % (spec, info['msg']))
+                self.module.fail_json(
+                    msg="Failure downloading %s, %s" % (spec, info['msg']),
+                    results=[]
```
Would you add trailing commas, please?

abadger replied:
Note.... this isn't a dictionary or list definition; it's a function call.  So the normal python style would be to not have a trailing comma....

abadger replied:
Out of curiosity, I went looking for the reason why the idiom is different for dictionaries and lists vs function calls and found it in PEP8:

> When trailing commas are redundant, they are often helpful when a version control system is used, when a list of values, arguments or imported items is expected to be extended over time.

Function calls are usually not extended arbitrarily later while dicts and lists of hardcoded values often are.  In the case of fail_json(), extra keyword arguments can be treated as more values to put into the results dict so it's not a strict contract but it is unlikely that we'd extend any given function call with more values.  So I'd lean towards not adding a trailing comma here.

webknjaz replied:
I normally add commas to not have to think about what it is. Also, when calling functions you don't know whether they'll be extended in future or know. It's one more thing to think about, so I just always automatically do this to multiline function args.

webknjaz commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,91 @@
+                module = self._templar.template("{{hostvars['%s']['ansible_facts']['pkg_mgr']}}" % self._task.delegate_to)
+            else:
+                module = self._templar.template("{{ansible_facts.pkg_mgr}}")
+        except Exception:
```
Maybe catch something more specific?

abadger replied:
I don't think the logic here is to catch something more specific.  But I always leave a comment in cases like these:
```
except Exception:
    # If module doesn't get set here, we fallback to running facts and trying again, below
    pass
```

webknjaz replied:
IMHO it's an antipattern when you catch everything wholesale. You mask future bugs this way.

maxamillion replied:
@webknjaz probably, I 100% just cribbed that from the package action plugin and didn't think much about the specific exception handler

abadger commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,91 @@
+                result.update(self._execute_module(module_name=module, module_args=new_module_args, task_vars=task_vars, wrap_async=self._task.async_val))
+                # Now fall through to cleanup
+        else:
+            result.update({'failed': True, 'msg': "Could not detect which major revision of yum is in use, which is required to determine module backend."})
```
Add that the user can specify the backend to use via the ```use``` parameter.

abadger replied:
Still need to add something like ```You can manually specify use_backend to tell the module whether to use the yum (yum-3) or dnf (yum-4) backend.```

abadger commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,91 @@
+        del tmp  # tmp no longer has any effect
+
+        # Carry-over concept from the package action plugin
+        module = "auto"
```
This doesn't appear to support the use parameter but probably should.

abadger commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,91 @@
+        if module not in ["yum", "yum4", "dnf"]:
+            facts = self._execute_module(module_name="setup", module_args=dict(filter="ansible_pkg_mgr", gather_subset="!all"), task_vars=task_vars)
+            display.debug("Facts %s" % facts)
+            module = facts.get("ansible_facts", {}).get("ansible_pkg_mgr", "auto")
```
When delegating, I'd think that we'd need to run the facts module for the delegated host rather than for the inventory_hostname host....

abadger replied:
@bcoca corrected me that _execute_module() takes care of delegation so this is fine.

After discussion, though, we did decide that we should set the ansible_pkg_mgr fact when we return from this action plugin so that we don't end up having to discover it every time the actin plugin is run on the same host.

abadger replied:
bcoca also mentioned that the facts end up being set on the host that's being processed in the task loop, not one that's being delegated_to.  So we should only return the ansible_pkg_mgr fact if we are not delegating.

abadger replied:
Still need to add the ansible_pkg-mgr fact to results if:
* We selected one of yum, yum4, or dnf
* And we are not delegating.

abadger replied:
https://docs.ansible.com/ansible/2.6/reference_appendices/common_return_values.html#ansible-facts

abadger replied:
So I guess that this is where we need to set the fact in results.  This is the only place where we know that we've had to run the facts module in order to retrieve the package manager.

abadger commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,91 @@
+                result.update({'failed': True, 'msg': "Could not find a yum module backend for %s." % module})
+            else:
+                # run either the yum (yum3) or dnf (yum4) backend module
+                new_module_args = self._task.args.copy()
```
If you don't modify the task args then you don't need to copy() them.  However, I think we want to add a ```use``` parameter for the action plugin and we will want to delete that parameter before we pass the args on to the module.  So this section would look like:
``` python
new_module_args = self._task.args.copy()
del new_module_args['use']
[...]
result.update(self._execute_module(module_name=module, module_args=new_module_args, task_vars=task_vars, wrap_async=self._task.async_val))
```

webknjaz commented:
@maxamillion you know you can just hide bot's comments instead of removing them?

abadger commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,94 @@
+                # run either the yum (yum3) or dnf (yum4) backend module
+                new_module_args = self._task.args.copy()
+                if 'use' in new_module_args:
+                    del new_module_args['use']
```
this got named use_backend

maxamillion replied:
correct, it did

abadger commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,101 @@
+                    module = self._templar.template("{{hostvars['%s']['ansible_facts']['pkg_mgr']}}" % self._task.delegate_to)
+                else:
+                    module = self._templar.template("{{ansible_facts.pkg_mgr}}")
+                    result['ansible_facts'] = {'pkg_mgr': self._templar.template('{{ansible_facts.pkg_mgr}}')}
```
As you've noticed, we don't want to do this here because in this case the pkg_mgr is coming from a pre-existing fact.

maxamillion commented:
This unintentionally also fixes https://github.com/ansible/ansible/issues/22407

abadger commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,102 @@
+        if module not in ["yum", "yum4", "dnf"]:
+            facts = self._execute_module(module_name="setup", module_args=dict(filter="ansible_pkg_mgr", gather_subset="!all"), task_vars=task_vars)
+            display.debug("Facts %s" % facts)
+            module = facts.get("ansible_facts", {}).get("ansible_pkg_mgr", "auto")
```
need one more conditional:
``` python
# Only set the pkg_mgr if we're not delegating
 if not self._task.delegate_to: 
    module = facts.get("ansible_facts", {}).get("ansible_pkg_mgr", "auto")
````

bcoca commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,102 @@
+
+        if module == 'auto':
+            try:
+                if self._task.delegate_to:  # if we delegate, we should use delegated host's facts
```
if we delegate_facts we should return them  `and not self._task.delegate_facts'

bcoca commented on lib/ansible/plugins/action/yum.py:
```diff
@@ -0,0 +1,102 @@
+            display.debug("Facts %s" % facts)
+            module = facts.get("ansible_facts", {}).get("ansible_pkg_mgr", "auto")
+            if not self._task.delegate_to and module in ('dnf', 'yum', 'yum4'):
+                result['ansible_facts'] = {'pkg_mgr': module}
```
we should also return if we both delegate executions and delegate_facts

maxamillion committed:
update changelog and action plugin based on feedback

acozine provided changelog feedback which was incorporated and bcoca
identified an issue with the yum action plugin not returning the
pkg_mgr fact if the task delegates facts

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
update yum action plugin docstring with acozine recommendations

Signed-off-by: Adam Miller <admiller@redhat.com>

abadger commented:
Merged to devel for inclusion in 2.7.0
